# Pet Shelter Analysis

Data Analyst TechDegree, Unit 5 Assignment<br>Kristen Bouzaid, 29 November 2022

In [22]:
import pandas as pd
intakes = pd.read_csv('aac_intakes.csv')
intakes_outcomes = pd.read_csv('aac_intakes_outcomes.csv')
outcomes = pd.read_csv('aac_outcomes.csv')

## Understanding the Data

In [26]:
# looking at the data to determine number of rows and columns
intakes.shape, outcomes.shape, intakes_outcomes.shape

((80187, 12), (80681, 12), (79672, 41))

In [27]:
# a different way of looking at number of rows
len(intakes), len(outcomes), len(intakes_outcomes)

(80187, 80681, 79672)

In [28]:
intakes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80187 entries, 0 to 80186
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   age_upon_intake   80187 non-null  object
 1   animal_id         80187 non-null  object
 2   animal_type       80187 non-null  object
 3   breed             80187 non-null  object
 4   color             80187 non-null  object
 5   datetime          80187 non-null  object
 6   datetime2         80187 non-null  object
 7   found_location    80187 non-null  object
 8   intake_condition  80187 non-null  object
 9   intake_type       80187 non-null  object
 10  name              55603 non-null  object
 11  sex_upon_intake   80186 non-null  object
dtypes: object(12)
memory usage: 7.3+ MB


In [29]:
outcomes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80681 entries, 0 to 80680
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   age_upon_outcome  80673 non-null  object
 1   animal_id         80681 non-null  object
 2   animal_type       80681 non-null  object
 3   breed             80681 non-null  object
 4   color             80681 non-null  object
 5   date_of_birth     80681 non-null  object
 6   datetime          80681 non-null  object
 7   monthyear         80681 non-null  object
 8   name              56116 non-null  object
 9   outcome_subtype   36893 non-null  object
 10  outcome_type      80667 non-null  object
 11  sex_upon_outcome  80679 non-null  object
dtypes: object(12)
memory usage: 7.4+ MB


In [30]:
intakes_outcomes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79672 entries, 0 to 79671
Data columns (total 41 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age_upon_outcome            79672 non-null  object 
 1   animal_id_outcome           79672 non-null  object 
 2   date_of_birth               79672 non-null  object 
 3   outcome_subtype             36348 non-null  object 
 4   outcome_type                79662 non-null  object 
 5   sex_upon_outcome            79671 non-null  object 
 6   age_upon_outcome_(days)     79672 non-null  int64  
 7   age_upon_outcome_(years)    79672 non-null  float64
 8   age_upon_outcome_age_group  79672 non-null  object 
 9   outcome_datetime            79672 non-null  object 
 10  outcome_month               79672 non-null  int64  
 11  outcome_year                79672 non-null  int64  
 12  outcome_monthyear           79672 non-null  object 
 13  outcome_weekday             796

In [31]:
intakes.head()

age_upon_intake animal_id animal_type  \
0         8 years   A706918         Dog   
1       11 months   A724273         Dog   
2         4 weeks   A665644         Cat   
3         4 years   A682524         Dog   
4         2 years   A743852         Dog   

                                   breed        color  \
0               English Springer Spaniel  White/Liver   
1                            Basenji Mix  Sable/White   
2                 Domestic Shorthair Mix       Calico   
3  Doberman Pinsch/Australian Cattle Dog     Tan/Gray   
4                 Labrador Retriever Mix    Chocolate   

                  datetime                datetime2  \
0  2015-07-05T12:59:00.000  2015-07-05T12:59:00.000   
1  2016-04-14T18:43:00.000  2016-04-14T18:43:00.000   
2  2013-10-21T07:59:00.000  2013-10-21T07:59:00.000   
3  2014-06-29T10:38:00.000  2014-06-29T10:38:00.000   
4  2017-02-18T12:46:00.000  2017-02-18T12:46:00.000   

                       found_location intake_condition      intake_type  \
0    9409 Bluegrass Dr in Austin (TX)           Normal            Stray   
1  2818 Palomino Trail in Austin (TX)           Normal            Stray   
2                         Austin (TX)             Sick            Stray   
3       800 Grove Blvd in Austin (TX)           Normal            Stray   
4                         Austin (TX)           Normal  Owner Surrender   

      name sex_upon_intake  
0    Belle   Spayed Female  
1  Runster     Intact Male  
2      NaN   Intact Female  
3      Rio   Neutered Male  
4     Odin   Neutered Male

In [32]:
outcomes.head()

age_upon_outcome animal_id animal_type                    breed  \
0          2 weeks   A684346         Cat   Domestic Shorthair Mix   
1           1 year   A666430         Dog               Beagle Mix   
2           1 year   A675708         Dog                 Pit Bull   
3          9 years   A680386         Dog  Miniature Schnauzer Mix   
4         5 months   A683115       Other                  Bat Mix   

          color        date_of_birth             datetime  \
0  Orange Tabby  2014-07-07T00:00:00  2014-07-22T16:04:00   
1   White/Brown  2012-11-06T00:00:00  2013-11-07T11:47:00   
2    Blue/White  2013-03-31T00:00:00  2014-06-03T14:20:00   
3         White  2005-06-02T00:00:00  2014-06-15T15:50:00   
4         Brown  2014-01-07T00:00:00  2014-07-07T14:04:00   

             monthyear     name outcome_subtype outcome_type sex_upon_outcome  
0  2014-07-22T16:04:00      NaN         Partner     Transfer      Intact Male  
1  2013-11-07T11:47:00     Lucy         Partner     Transfer    Spayed Female  
2  2014-06-03T14:20:00  *Johnny             NaN     Adoption    Neutered Male  
3  2014-06-15T15:50:00   Monday         Partner     Transfer    Neutered Male  
4  2014-07-07T14:04:00      NaN     Rabies Risk   Euthanasia          Unknown

In [33]:
intakes_outcomes.head()

age_upon_outcome animal_id_outcome        date_of_birth outcome_subtype  \
0         10 years           A006100  2007-07-09 00:00:00             NaN   
1          7 years           A006100  2007-07-09 00:00:00             NaN   
2          6 years           A006100  2007-07-09 00:00:00             NaN   
3         10 years           A047759  2004-04-02 00:00:00         Partner   
4         16 years           A134067  1997-10-16 00:00:00             NaN   

      outcome_type sex_upon_outcome  age_upon_outcome_(days)  \
0  Return to Owner    Neutered Male                     3650   
1  Return to Owner    Neutered Male                     2555   
2  Return to Owner    Neutered Male                     2190   
3         Transfer    Neutered Male                     3650   
4  Return to Owner    Neutered Male                     5840   

   age_upon_outcome_(years) age_upon_outcome_age_group     outcome_datetime  \
0                      10.0                (7.5, 10.0]  2017-12-07 14:07:00   
1                       7.0                 (5.0, 7.5]  2014-12-20 16:35:00   
2                       6.0                 (5.0, 7.5]  2014-03-08 17:10:00   
3                      10.0                (7.5, 10.0]  2014-04-07 15:12:00   
4                      16.0               (15.0, 17.5]  2013-11-16 11:54:00   

   ...  age_upon_intake_age_group      intake_datetime intake_month  \
0  ...                (7.5, 10.0]  2017-12-07 00:00:00           12   
1  ...                 (5.0, 7.5]  2014-12-19 10:21:00           12   
2  ...                 (5.0, 7.5]  2014-03-07 14:26:00            3   
3  ...                (7.5, 10.0]  2014-04-02 15:55:00            4   
4  ...               (15.0, 17.5]  2013-11-16 09:02:00           11   

  intake_year  intake_monthyear  intake_weekday  intake_hour  intake_number  \
0        2017           2017-12        Thursday           14            1.0   
1        2014           2014-12          Friday           10            2.0   
2        2014           2014-03          Friday           14            3.0   
3        2014           2014-04       Wednesday           15            1.0   
4        2013           2013-11        Saturday            9            1.0   

             time_in_shelter time_in_shelter_days  
0  0 days 14:07:00.000000000             0.588194  
1  1 days 06:14:00.000000000             1.259722  
2  1 days 02:44:00.000000000             1.113889  
3  4 days 23:17:00.000000000             4.970139  
4  0 days 02:52:00.000000000             0.119444  

[5 rows x 41 columns]

In [34]:
# check to see if there are any instances in intakes_outcomes where animal_id_intake and animal_id_outcome differ, just to understand the data a bit better
intakes_outcomes[intakes_outcomes['animal_id_outcome']!=intakes_outcomes['animal_id_intake']].size

0

In [36]:
# check which items have missing data
intakes.count()

age_upon_intake     80187
animal_id           80187
animal_type         80187
breed               80187
color               80187
datetime            80187
datetime2           80187
found_location      80187
intake_condition    80187
intake_type         80187
name                55603
sex_upon_intake     80186
dtype: int64

In [38]:
# out of curiosity, check for entry with missing data in sex_upon_intake
intakes[intakes.sex_upon_intake.isna()].head()

age_upon_intake animal_id animal_type      breed        color  \
13715         7 years   A667395         Dog  Dachshund  Brown Merle   

                      datetime                datetime2     found_location  \
13715  2013-11-17T13:15:00.000  2013-11-17T13:15:00.000  Pflugerville (TX)   

      intake_condition      intake_type   name sex_upon_intake  
13715           Normal  Owner Surrender  Diego             NaN

In [41]:
outcomes.count()

age_upon_outcome    80673
animal_id           80681
animal_type         80681
breed               80681
color               80681
date_of_birth       80681
datetime            80681
monthyear           80681
name                56116
outcome_subtype     36893
outcome_type        80667
sex_upon_outcome    80679
dtype: int64

In [42]:
intakes_outcomes.count()

age_upon_outcome              79672
animal_id_outcome             79672
date_of_birth                 79672
outcome_subtype               36348
outcome_type                  79662
sex_upon_outcome              79671
age_upon_outcome_(days)       79672
age_upon_outcome_(years)      79672
age_upon_outcome_age_group    79672
outcome_datetime              79672
outcome_month                 79672
outcome_year                  79672
outcome_monthyear             79672
outcome_weekday               79672
outcome_hour                  79672
outcome_number                79672
dob_year                      79672
dob_month                     79672
dob_monthyear                 79672
age_upon_intake               79672
animal_id_intake              79672
animal_type                   79672
breed                         79672
color                         79672
found_location                79672
intake_condition              79672
intake_type                   79672
sex_upon_intake             

## Cleaning the Data

In [43]:
# remove duplicates from each dataset
intakes.drop_duplicates(keep='last', inplace=True)
outcomes.drop_duplicates(keep='last', inplace=True)
intakes_outcomes.drop_duplicates(keep='last', inplace=True)

In [44]:
# check to see how many entries remain
#len(intakes_unique), len(outcomes_unique), len(intakes_outcomes_unique)
len(intakes), len(outcomes), len(intakes_outcomes)

(80161, 80671, 79637)

In [46]:
# convert datetime fields to datetime data type
intakes['datetime'] = pd.to_datetime(intakes['datetime'])
intakes['datetime2'] = pd.to_datetime(intakes['datetime2'])
intakes.dtypes

age_upon_intake             object
animal_id                   object
animal_type                 object
breed                       object
color                       object
datetime            datetime64[ns]
datetime2           datetime64[ns]
found_location              object
intake_condition            object
intake_type                 object
name                        object
sex_upon_intake             object
dtype: object

In [48]:
#check to see if there datetime and datetime2 in intakes are identical
(intakes['datetime']-intakes['datetime2']).sum()

Timedelta('0 days 00:00:00')

In [53]:
# separate the date and add new columns for month and year to allow calculations
intakes['month'] = intakes['datetime'].dt.month
intakes['year'] = intakes['datetime'].dt.year
#check
intakes.head(2)

age_upon_intake animal_id animal_type                     breed  \
0         8 years   A706918         Dog  English Springer Spaniel   
1       11 months   A724273         Dog               Basenji Mix   

         color            datetime           datetime2  \
0  White/Liver 2015-07-05 12:59:00 2015-07-05 12:59:00   
1  Sable/White 2016-04-14 18:43:00 2016-04-14 18:43:00   

                       found_location intake_condition intake_type     name  \
0    9409 Bluegrass Dr in Austin (TX)           Normal       Stray    Belle   
1  2818 Palomino Trail in Austin (TX)           Normal       Stray  Runster   

  sex_upon_intake  month  year  
0   Spayed Female      7  2015  
1     Intact Male      4  2016

In [51]:
# assign value for entry with missing data in sex_upon_intake
intakes['sex_upon_intake'] = intakes['sex_upon_intake'].fillna('Unknown')
intakes_outcomes['sex_upon_intake'] = intakes_outcomes['sex_upon_intake'].fillna('Unknown')

In [55]:
#outcomes['age_upon_outcome'] = outcomes['age_upon_outcome'].astype('int')

## 1. Is there an area where more pets are found?
Find the top 5 places where animals are found so the shelter can coordiante with local volunteers and animal control to monitor these areas. Hint: Try to group the found locations together.

Approach: For this question, I grouped the intakes data by 'found_location', determined the counts based on 'found_location', sorted by the counts to determine locations with most pickups, and then took the top 5.

In [56]:
intakes.groupby('found_location').size().reset_index(name='pickupcount').sort_values(['pickupcount'],ascending=False).head(5)

found_location  pickupcount
25970                        Austin (TX)        14437
32692               Outside Jurisdiction          948
35512                        Travis (TX)          921
21133  7201 Levander Loop in Austin (TX)          517
27901                     Del Valle (TX)          411

The data needs to be specific enough that local volunteers can target an area, so 'Austin' and 'Outside Jurisdiction' are too broad. To focus on found locations with street numbers, I'll search for entries containing a number. Chances are if there are many pick ups at one address, then adjacent areas will also be of interest.

In [57]:
intakes[intakes['found_location'].str.contains('1|2|3|4|5|6|7|8|9|0')].groupby('found_location').size().reset_index(name='pickupcount').sort_values(['pickupcount'],ascending=False).head(5)

found_location  pickupcount
21132   7201 Levander Loop in Austin (TX)          517
14955    4434 Frontier Trl in Austin (TX)          163
3960     124 W Anderson Ln in Austin (TX)          153
2519   1156 W Cesar Chavez in Austin (TX)           98
3315   12034 Research Blvd in Austin (TX)           98

### Final Answer: 
The five most common pick up locations that are specific enough that the shelter would be able to deploy volunteers to a certain area are listed above. 

## 2. What is the average number of pets found in a month in the year 2015? Are there months where there is a higher number of animals found?
Approach 1: Using intakes, I will filter the datetime field for 2015 and create a new dataframe called tfintakes (twenty-fifteen intakes). Then I will sum up animals taken in and divide by 12. Then I will group them by month and sort descending. I determined (above) that datetime and datetime1 are identical, so I will use datetime.


In [58]:
#sort by dates so only intakes from 2015 are returned
tfintakes = intakes[(intakes['datetime'] >= pd.Timestamp(2015,1,1)) & (intakes['datetime'] <= pd.Timestamp(2015,12,31))]
#check the length to make sure the answer is reasonable
len(tfintakes)
# consider adding index reset to tfintakes

18676

In [61]:
# to calculate average intakes per month, divide number of intakes in 2015 by 12 months, rounding to a whole number 
#      using int to round down to the closest whole number because 0.4 of a dog is a frightening thought!
int(len(tfintakes)/12)

1556

In [62]:
# group the 2015 intakes by month in descending order
tfintakes_by_month = tfintakes.groupby('month')
tfintakes_by_month.size().sort_values(ascending=False)

month
6     2189
5     2094
10    1740
8     1718
7     1635
9     1591
4     1543
11    1411
3     1345
1     1197
2     1118
12    1095
dtype: int64

### Final Answer: 
The average number of pets found in a month in the year 2015 is 1,556. How sad. :(  
The months with the most intakes (in descending order) were: June, May, October, August, July, September. The colder months (November, March, January, February, and December [in descending order]) had fewer intakes.

## 3. What is the ratio of incoming pets vs. adopted pets?
This key metric helps the shelter know how they are doing.

Approach: For this, I will compare the total number of intakes to the total number of adoptions. For this I will use the intakes_outcomes dataset. Question: Is an Adoption/Foster a true success? Perhaps look at that statistic as well, to only identify the true permanent adoptions.

In [63]:
# Total number of intakes
len(intakes)

80161

In [64]:
# a look at the different categories of outcome_type
intakes_outcomes.groupby('outcome_type').size()

outcome_type
Adoption           33583
Died                 690
Disposal             301
Euthanasia          6238
Missing               46
Relocate              15
Return to Owner    14784
Rto-Adopt            179
Transfer           23791
dtype: int64

In [65]:
# a look at the different categories of outcome_subtype
intakes_outcomes.groupby('outcome_subtype').size()

outcome_subtype
Aggressive               497
At Vet                    71
Barn                       3
Behavior                 133
Court/Investigation       23
Enroute                   49
Foster                  5489
In Foster                177
In Kennel                351
In Surgery                17
Medical                  265
Offsite                  350
Partner                19836
Possible Theft             9
Rabies Risk             2536
SCRP                    3202
Snr                      751
Suffering               2546
Underage                  28
dtype: int64

In [66]:
adoptions = intakes_outcomes[intakes_outcomes['outcome_type'] == "Adoption"]
adoptions.head(5)

age_upon_outcome animal_id_outcome        date_of_birth outcome_subtype  \
12         16 years           A200922  1997-10-03 00:00:00          Foster   
14         17 years           A210457  1999-06-01 00:00:00          Foster   
22         15 years           A226069  2000-06-17 00:00:00          Foster   
32         15 years           A249087  2001-01-02 00:00:00          Foster   
48         11 years           A274546  2002-05-25 00:00:00             NaN   

   outcome_type sex_upon_outcome  age_upon_outcome_(days)  \
12     Adoption    Neutered Male                     5840   
14     Adoption    Neutered Male                     6205   
22     Adoption    Neutered Male                     5475   
32     Adoption    Spayed Female                     5475   
48     Adoption    Spayed Female                     4015   

    age_upon_outcome_(years) age_upon_outcome_age_group     outcome_datetime  \
12                      16.0               (15.0, 17.5]  2013-11-22 09:44:00   
14                      17.0               (15.0, 17.5]  2016-10-07 12:34:00   
22                      15.0               (12.5, 15.0]  2015-10-29 15:00:00   
32                      15.0               (12.5, 15.0]  2016-11-15 10:37:00   
48                      11.0               (10.0, 12.5]  2013-11-16 12:24:00   

    ...  age_upon_intake_age_group      intake_datetime intake_month  \
12  ...               (15.0, 17.5]  2013-10-03 15:47:00           10   
14  ...               (15.0, 17.5]  2016-09-28 12:05:00            9   
22  ...               (12.5, 15.0]  2015-10-06 12:29:00           10   
32  ...               (12.5, 15.0]  2016-08-17 14:05:00            8   
48  ...               (10.0, 12.5]  2013-11-13 18:16:00           11   

   intake_year  intake_monthyear  intake_weekday  intake_hour  intake_number  \
12        2013           2013-10        Thursday           15            1.0   
14        2016           2016-09       Wednesday           12            1.0   
22        2015           2015-10         Tuesday           12            1.0   
32        2016           2016-08       Wednesday           14            1.0   
48        2013           2013-11       Wednesday           18            1.0   

               time_in_shelter time_in_shelter_days  
12  49 days 17:57:00.000000000            49.747917  
14   9 days 00:29:00.000000000             9.020139  
22  23 days 02:31:00.000000000            23.104861  
32  89 days 20:32:00.000000000            89.855556  
48   2 days 18:08:00.000000000             2.755556  

[5 rows x 41 columns]

In [67]:
len(adoptions)

33583

In [71]:
ratio_adoptions = len(intakes)/len(adoptions)
# round figure to two decimal places
round(ratio_adoptions, 2)

2.39

In [73]:
# adoptions_4eva is a look at how many are 'permanent' adoptions and not of the Foster subtype
adoptions_4eva = intakes_outcomes[(intakes_outcomes['outcome_type'] == "Adoption") & (intakes_outcomes['outcome_subtype'] != "Foster")]
len(adoptions_4eva)

28094

In [74]:
ratio_adoptions_4eva = len(intakes)/len(adoptions_4eva)
# round figure to two decimal places
round(ratio_adoptions_4eva, 2)

2.85

### Final Answer
For every 2.39 pets coming into the shelter, there is 1 'Adoption' outcome.
For every 2.85 pets coming into the shelter, there is 1 'Adoption' that is NOT a temporary foster situation.

## 4. What is the distribution of the types of animals in the shelter?
Find the count of each type of animal in the shelter.

Approach: Using the intakes dataset, group the animals by type and then find the size to determine counts, sorting in descending order to determine numbers of each type of animal.

In [75]:
intakes.groupby('animal_type').size().sort_values(ascending=False)

animal_type
Dog          45730
Cat          29651
Other         4429
Bird           342
Livestock        9
dtype: int64

### Final Answer: 
Dogs were the most common shelter residents at 45,743; cats were the second most common resident at 29,659. Other animals (wonder what these are...) came in at 4,434; there have been 342 bird residents, and livestock were the least common residents at only 9.

## 5. What are the adoption rate for specific breeds?
Find the top 5 dog breeds in the shelter (based on intake count) and then find the adoption percentage for each breed (based on number of Adoption outcomes for that breed).

Approach: First, I will filter the intakes by animal type and then sort by breed count to find the top 5 breeds that are taken in. Then, I will determine the number of adoptions by dog breed. I will combine these dataframes to view the number of adoptions for the top 5 breeds taken in, and then - in this new dataframe - display the calculation of adoption rate for each breed.

In [80]:
# Top 5 breeds of dog taken in
dog_intakes = intakes[intakes['animal_type'] == "Dog"].groupby('breed').size().reset_index(name='Number of Intakes').sort_values(by='Number of Intakes', ascending=False).head(5)
dog_intakes
dog_intakesdf = pd.DataFrame(dog_intakes).reset_index()
dog_intakesdf = dog_intakesdf.drop('index', axis=1)
dog_intakesdf

#intakes.groupby('found_location').size().reset_index(name='pickupcount').sort_values(['pickupcount'],ascending=False).head(5)

breed  Number of Intakes
0               Pit Bull Mix               6379
1    Chihuahua Shorthair Mix               4860
2     Labrador Retriever Mix               4840
3        German Shepherd Mix               1963
4  Australian Cattle Dog Mix               1105

In [81]:
# Number of adoptions per dog breed
dog_adoptions_by_breed = outcomes[(outcomes['animal_type'] == "Dog") & (outcomes['outcome_type'] == "Adoption")].groupby('breed').size().reset_index(name='Number of Adoptions').sort_values(by='Number of Adoptions', ascending=False).head(10)
dog_adoptions_by_breeddf = pd.DataFrame(dog_adoptions_by_breed).reset_index()
dog_adoptions_by_breeddf = dog_adoptions_by_breeddf.drop('index', axis=1)
dog_adoptions_by_breeddf

breed  Number of Adoptions
0     Labrador Retriever Mix                 2404
1               Pit Bull Mix                 2382
2    Chihuahua Shorthair Mix                 2293
3        German Shepherd Mix                  937
4  Australian Cattle Dog Mix                  619
5              Dachshund Mix                  394
6          Border Collie Mix                  355
7                  Boxer Mix                  313
8       Miniature Poodle Mix                  276
9              Catahoula Mix                  253

In [83]:
#Merging dataframes and calculating Adoption Rate as a percentage
adoption_top_five = dog_intakesdf.merge(
    dog_adoptions_by_breeddf,
    left_on=['breed'],
    right_on=['breed']
)
adoption_top_five["Adoption Rate %"] = ((adoption_top_five['Number of Adoptions']/adoption_top_five['Number of Intakes'])*100).round(2)

adoption_top_five.head()

breed  Number of Intakes  Number of Adoptions  \
0               Pit Bull Mix               6379                 2382   
1    Chihuahua Shorthair Mix               4860                 2293   
2     Labrador Retriever Mix               4840                 2404   
3        German Shepherd Mix               1963                  937   
4  Australian Cattle Dog Mix               1105                  619   

   Adoption Rate %  
0            37.34  
1            47.18  
2            49.67  
3            47.73  
4            56.02

### Final Answer
The top five dog breeds taken into the shelter, from most intakes to least, are: Pit Bull Mixes, Chihuahua Shorthair Mixes, Labrador Retriever Mixes, German Shepherd Mixes, and Australian Cattle Dog Mixes. Their adoption rates are: 37.34%, 47.18%, 49.67%, 47.73%, and 56.02%, respectively. I would ALSO be most likely to adopt an Australian Cattle Hound Mix, so those stats make sense to me.

## 6. What are the adoption rates for different colorings?
Find the top 5 colorings in the shelter (based on count) and then find the adoption percentage of each color.

Approach: I will assume that this questions applies to dogs (as per question 5). So first, in intakes, I will filter by animal_type for dogs, then filter intakes by coloring and then sort by color and count to find the top 5 colors. Then, I will calculate the # of adoptions for each color (using a similar technique as above), will merge the dataframes and calculate adoption percentages.

In [84]:
dog_intakes = intakes[intakes['animal_type'] == "Dog"].groupby('color').size().reset_index(name='Number of Intakes').sort_values(by='Number of Intakes', ascending=False).head(5)
dog_intakes
dog_intakesdf = pd.DataFrame(dog_intakes).reset_index()
dog_intakesdf = dog_intakesdf.drop('index', axis=1)
dog_intakesdf

#intakes.groupby('found_location').size().reset_index(name='pickupcount').sort_values(['pickupcount'],ascending=False).head(5)

color  Number of Intakes
0  Black/White               5253
1        Black               2527
2  Brown/White               2500
3    Tan/White               2473
4        White               2378

In [85]:
dog_adoptions_by_breed = outcomes[(outcomes['animal_type'] == "Dog") & (outcomes['outcome_type'] == "Adoption")].groupby('color').size().reset_index(name='Number of Adoptions').sort_values(by='Number of Adoptions', ascending=False).head(10)
dog_adoptions_by_breeddf = pd.DataFrame(dog_adoptions_by_breed).reset_index()
dog_adoptions_by_breeddf = dog_adoptions_by_breeddf.drop('index', axis=1)
dog_adoptions_by_breeddf

color  Number of Adoptions
0  Black/White                 2609
1    Tan/White                 1178
2  Brown/White                 1123
3        Black                 1089
4     Tricolor                  969
5          Tan                  938
6        White                  932
7    Black/Tan                  873
8        Brown                  790
9  White/Black                  739

In [86]:
#Merging dataframes
adoption_top_five = dog_intakesdf.merge(
    dog_adoptions_by_breeddf,
    left_on=['color'],
    right_on=['color']
)
adoption_top_five["Adoption Rate %"] = ((adoption_top_five['Number of Adoptions']/adoption_top_five['Number of Intakes'])*100).round(2)

adoption_top_five.head()

color  Number of Intakes  Number of Adoptions  Adoption Rate %
0  Black/White               5253                 2609            49.67
1        Black               2527                 1089            43.09
2  Brown/White               2500                 1123            44.92
3    Tan/White               2473                 1178            47.63
4        White               2378                  932            39.19

### Final Answer
The top five dog colors taken into the shelter, from most intakes to least, are: Black/White, Black, Brown/White, Tan/White, and White. Their adoption rates are: 49.67%, 43.09%, 44.92%, 47.63%, and 39.19%, respectively.

## 7. About how many animals are spayed/neutered each month?
This will help the shelter allocate resources and staff. Assume that all intact males and females will be spayed/neutered. 

Approach: The assumption that all intact males and females will be spayed/neutered assumes that even animals returned to rheir owners will be spayed/neutered, which I think is incorrect. However, I will approach this two ways:

1. Assume ALL intact intakes are neutered, regardless of outcome_type. This assumes that either they are neutered in the same month they arrive or there is a consistent stream of neuters so the variance between intake date and neuter date is negligible.

2. Determine number of outcomes in which sex_upon_outtake was "Neutered Male" or "Neutered Female" and sex_upon_intake is DIFFERENT than sex_upon_outcome. This positively identifies cases in which a dogs sexual status has changed, rather than making assumptions based on intakes status.

In [87]:
# Approach 1
# determine numbers of intact intakes
intact_intakes = intakes[intakes['sex_upon_intake'].str.contains('Intact')]
intact_intakes_by_month = intact_intakes.groupby(['year','month']).size().sum()
round(intact_intakes_by_month,2)

49337

In [88]:
intact_intakes_by_month = intact_intakes.groupby(['year','month']).size().mean()
round(intact_intakes_by_month)

914

In [90]:
# Approach 2
# determine number of sex changes
intact_outcomes = intakes_outcomes[intakes_outcomes['sex_upon_intake'].str.contains('Intact')]
neuters = intact_outcomes[intact_outcomes['sex_upon_outcome'] != intact_outcomes['sex_upon_intake']]
len(neuters)

29978

In [91]:
neuters.groupby('sex_upon_outcome').size().sort_values(ascending=False).head()

sex_upon_outcome
Neutered Male    15581
Spayed Female    14397
dtype: int64

In [92]:
neuters_by_month = neuters.groupby(['intake_year','intake_month']).size().mean()
round(neuters_by_month)

555

### Final Answer: 
Approach 1: There are approximately 914 neuters conducted each month.
Approach 2: There are approximately 555 neuters conducted each month.

## Extra Credit

## 1. How many animals in the shelter are repeats? Which animal was returned to the shelter the most?
This means the animal has been brought in more than once.

Approach: For this, I will use the intakes dataframe. I will assume that each time a 'repeat customer' is taken into the shelter, it is given the same animal id. However, for each intake occurrence, it will have a new unique index. To answer the first question, I will count up the number of animal_ids that appear more than once in the intakes dataframe. 

To answer the second part of the question, I will get details on the animal with the max number of intakes.

In [93]:
intakes_per_animal_id = intakes.groupby('animal_id').size().reset_index(name='Number of Intakes').sort_values(by='Number of Intakes',ascending=False)
repeat_intakes = intakes_per_animal_id[intakes_per_animal_id['Number of Intakes'] > 1]
len(repeat_intakes)

6133

In [94]:
repeat_intakes.head(1)

animal_id  Number of Intakes
41352   A721033                 13

### Final Answer: 
There were 6,133 'repeat offenders', i.e. animals who had been admitted more than once. The animal that was admitted most was animal_id A721033, with 13 intakes.

## 2. What are the adoption rates for the following age groups?
* baby: 4 months and less
* young: 5 months - 2 years
* adult: 3 years - 10 years
* senior: 11+

Approach: I am first going to add a field to the intakes_outcomes table to indicate what age bracket the animal falls into. I will base this on age_upon_outcomes(days). Then I will create a new dataframe to house by data. To collect and calculate the data, I will use a for loop to cycle through the age group label list and calculate the adoption rate data. This will then be added to the dataframe.

In [97]:
# create data bins according to age in days; 15000 days is used as the oldest age - perhaps for a very senior parrot.
bins = [0,120,730,3650,15000]    
labels = ['baby', 'young', 'adult', 'senior']

# add an age_group field to the intakes_outcomes dataset
intakes_outcomes['age_group'] = pd.cut(intakes_outcomes['age_upon_outcome_(days)'], bins=bins, labels=labels, right=True)

# create a new dataframe shell for the adoption data
# for each age group, calculate the total number of animals in that age group using len(), the number of adoptions using outcome_type, and the % adopted
# add each record to the dataframe, looping over the labels list

adoption_age_data = pd.DataFrame([], columns=['Type','Total_Number', 'Total_Number_of_Adoptions', 'AdoptionPerc'])

for item in labels:
    age_category = item
    number_in_age_group = len(intakes_outcomes[intakes_outcomes['age_group'] == item])
    number_adopted = len(intakes_outcomes[(intakes_outcomes['age_group'] == item) & (intakes_outcomes['outcome_type'] == "Adoption")])
    percent_adopted = round((number_adopted/number_in_age_group*100),2)
    record = {'Type':[age_category], 'Total_Number':[number_in_age_group], 'Total_Number_of_Adoptions':[number_adopted], 'AdoptionPerc':[percent_adopted]}
    recorddf=pd.DataFrame(record)
    adoption_age_data = pd.concat([adoption_age_data,recorddf],ignore_index=True)
adoption_age_data

Type Total_Number Total_Number_of_Adoptions  AdoptionPerc
0    baby        24518                     12591         51.35
1   young        34776                     14507         41.72
2   adult        18099                      6030         33.32
3  senior         2150                       452         21.02

### Final Answer: 
The adoption rates for baby animals is 51.35%, for young animals is 41.72%, for adult animals is 33.32%, and for senior animals is 21.02%

## 3. If spay/neuter for a dog costs \\$100 and a spay/neuter for a cat costs \\$50, how much did the shelter spend in 2015 on these procedures?

Approach: I will assume that all intact intakes in 2015 were spayed or neutered, unless they were returned to their owner.

In [105]:
# intact_intakes from Question 7 is the subset from intakes of animals with the work 'Intact' in their sex_upon_intake
# tfintact_intakes (below) are the intact intakes in 2015 (tf)
tfintact_intakes = intact_intakes[(intact_intakes['datetime'] >= pd.Timestamp(2015,1,1)) & (intact_intakes['datetime'] <= pd.Timestamp(2015,12,31))]
tfcat_neuters = len(tfintact_intakes[tfintact_intakes['animal_type'] == 'Cat'])
tfcat_neuters

5061

In [106]:
tfdog_neuters = len(tfintact_intakes[tfintact_intakes['animal_type'] == 'Dog'])
tfdog_neuters

6086

In [107]:
cost_of_2015_neuters = (tfcat_neuters * 50) + (tfdog_neuters * 100)
print(cost_of_2015_neuters)

861650


### Final Answer: 
In 2015, the shelter spent \\$861,650 to spay or neuter 5,061 cats and 6,086 dogs.

## 4. Describe your findings and methods for each of the above quesitons. Use markdown cells in JupyterLab to detail. 
How did you get your answer?
What was your final answer?

### Final Answer: See above for approaches and final answers for all questions